csharp/src/Drivers/BigQuery/RetryManager.cs

```diff
@@ -0,0 +1,90 @@
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Adbc.Drivers.BigQuery
+{
+    /// <summary>
+    /// Class that will retry calling a method with an exponential backoff.
+    /// </summary>
+    class RetryManager
+    {
+        public static async Task<T> ExecuteWithRetriesAsync<T>(
+           Func<Task<T>> action,
+           int maxRetries = 5,
+           int initialDelayMilliseconds = 200)
+        {
+            return await ExecuteWithRetriesAsync<T>(null, action, maxRetries, initialDelayMilliseconds);
+        }
+
+        public static async Task<T> ExecuteWithRetriesAsync<T>(
+            ITokenProtectedResource? tokenProtectedResource,
+            Func<Task<T>> action,
+            int maxRetries = 5,
+            int initialDelayMilliseconds = 200)
+        {
+            if (action == null)
+            {
+                throw new AdbcException("There is no method to retry", AdbcStatusCode.InvalidArgument);
+            }
+
+            int retryCount = 0;
+            int delay = initialDelayMilliseconds;
+
+            while (retryCount < maxRetries)
+            {
+                try
+                {
+                    T result = await action();
+                    return result;
+                }
+                catch (Exception ex)
+                {
+                    retryCount++;
+                    if (retryCount >= maxRetries)
+                    {
+                        if ((tokenProtectedResource?.UpdateToken != null))
+                        {
+                            if (tokenProtectedResource?.TokenRequiresUpdate(ex) == true)
+                            {
+                                throw new AdbcException($"Cannot update access token after {maxRetries} tries", AdbcStatusCode.Unauthenticated, ex);
+                            }
+                        }
+
+                        throw new AdbcException($"Cannot execute {action.Method.Name} after {maxRetries} tries", AdbcStatusCode.UnknownError, ex);
+                    }
+
+                    if ((tokenProtectedResource?.UpdateToken != null))
+                    {
+                        if (tokenProtectedResource.TokenRequiresUpdate(ex) == true)
+                        {
+                            await tokenProtectedResource.UpdateToken();
+                        }
+                    }
+
+                    await Task.Delay(delay);
+                    delay *= 2;
```

Comment:
I think this should stop doubling at some point. On the tenth iteration, we're waiting nearly 3.5 minutes. If it finishes around thirty seconds into the eleventh wait, then we still have to wait almost three more minutes before checking again.

Reply:
I'm not sure whether the new code `delay += delay` is supposed to be different than `delay *= 2`. Did you mean `delay += initialDelayMilliseconds`? Can I suggest an alternative of something like `delay = Math.Min(2 * delay, 5000)`?